src/NServiceBus.RavenDB/Timeouts/QueryTimeouts.cs

```diff
@@ -21,93 +24,85 @@ public QueryTimeouts(IDocumentStore documentStore, string endpointName)
 
-                do
+                var qhi = new Reference<QueryHeaderInformation>();
+                using (var enumerator = await session.Advanced.StreamAsync(query, qhi).ConfigureAwait(false))
                 {
-                    if (CancellationRequested())
+                    while (await enumerator.MoveNextAsync().ConfigureAwait(false))
                     {
-                        return new TimeoutsChunk(EmptyTimeouts, nextTimeToRunQuery);
-                    }
-
-                    // This is all an unexecuted Raven query expression
-                    var query = GetChunkQuery(session)
-                        .Statistics(out statistics)
-                        .Where(t => t.Time >= startSlice && t.Time <= now)
-                        .Skip(skipCount)
-                        .Take(maximumPageSize)
-                        .Select(to => new { to.Id, to.Time }); // Must be anonymous type so Raven server can understand
-
-                    var dueTimeouts = await query.ToListAsync().ConfigureAwait(false);
+                        if (CancellationRequested())
+                        {
+                            return new TimeoutsChunk(EmptyTimeouts, nextTimeoutToExpire);
+                        }
 
-                    results.AddRange(dueTimeouts.Select(t => new TimeoutsChunk.Timeout(t.Id, t.Time)));
-
-                    if (CancellationRequested())
-                    {
-                        return new TimeoutsChunk(EmptyTimeouts, nextTimeToRunQuery);
+                        results.Add(new TimeoutsChunk.Timeout(enumerator.Current.Document.Id, enumerator.Current.Document.Time));
                     }
-
-                    skipCount = results.Count + statistics.SkippedResults;
-                    totalCount = statistics.TotalResults;
                 }
-                while (results.Count < totalCount);
 
-                var nextDueTimeout = await
-                    GetChunkQuery(session)
+                var nextTimeout = await GetChunkQuery(session)
                         .Where(t => t.Time > now)
-                        .Select(t => t.Time)
+                        .Take(1)
+                        .Select(to => new { to.Time }) // Must be anonymous type so Raven server can understand
                         .FirstOrDefaultAsync()
                         .ConfigureAwait(false);
 
-                if (nextDueTimeout != default(DateTime))
+                if (nextTimeout != null)
                 {
-                    nextTimeToRunQuery = nextDueTimeout;
+                    nextTimeoutToExpire = nextTimeout.Time;
                 }
-            }
 
-            if (statistics.IsStale && results.Count == 0)
-            {
-                nextTimeToRunQuery = now;
+                // Next execution is either now if we know we got stale results or at the start of the next chunk, otherwise we delay the next execution a bit
+                else if (qhi.Value.IsStale && results.Count == 0)
```

Comment:
What if the index is stale but we got some results? Shouldn't we also start from "now" then too?